### ISSUE TITLE
SDK switcher is not working properly with Oculus and Vive?

### ISSUE THREAD

**Issue by guiglass** [thestonefox/VRTK#1642]
I can not get the SDK switcher to work with both my Rift cv1 and Vive headsets. I must change the SDK order then save the project and reopen it to use the other headset. I can make both the Oculus and Vive work in the same project, but I can not switch between them in a single session, however I can create two projects and have both headsets running at the same time. I do not suspect GFX or USB bottlenecks.

### Environment

 * Source of VRTK is Github.
 * Version of Unity3D 2017.1.03f
 * Hardware used Vive and Oculus
 * SDK used OpenVR and SteamVR and Oculus Utilities

### Steps to reproduce
* With both Vive and Oculus powered on and manager apps running.
* Create a new project.
* Import OpenVR, SteamVR and Oculus Utilities.
* Accept and install OpenVR 1.20.
* Load any of the VRTK examples.
* Press play,

  > If OuclusVR is highest in list of Setups upon project loading, then Rift will be rendered to.
  > If SteamVR is highest in list of Setups upon project loading, then Vive will be rendered to.

* Using the SDK switcher, change from Rift to Vive (or vice versa).
* At this point I notice the head and hand tracking does change to the proper headset.
* But the Oculus shows a message saying "Unity.exe is taking too long" or SteamVR reports "Unresponsive Unity.exe".

The only way to switch between them now is to put the desired SDK as the highest entry in the list of Setups, then restart Unity, and most of the time the other headset will be picked up properly.

### Expected behavior
When the SKD is switched the proper headset should start rendering the scene (no error messages).

I have two GTX 1070 cards and each headset is pluged into each card's HDMI port. Also I have a PCIE USB-3 card so (among the many USB ports configurations I have tried) I do have the Vive and Rift on separate USB3.0 controllers.
I really require some assistance with this, I have been working on trying to find a solution for a few solid days now!

**Comment by bddckr** [comment]
This is a bug/issue with the way SteamVR's Oculus driver or the Oculus runtime works.

Oculus' runtime just sees SteamVR as a rendering app, it doesn't matter that SteamVR actually let's multiple apps "render to itself" which then shows up on the Rift.

Once you switch between SteamVR/OpenVR and Oculus' SDK at runtime you'll confuse the Oculus runtime as it will wait for the rendering to finally start, but it won't ever since SteamVR isn't correctly connected to the Oculus runtime anymore. The latter can only be fixed by restarting SteamVR in my experience: It will then tell the Oculus runtime again about itself, thus the Oculus runtime thinks SteamVR is the rendering app and everything works.

---

Also note that you can run the Rift through SteamVR just fine, you just have to enforce the Oculus driver to be used in OpenVR if you have both connected (SteamVR prefers the Vive when it finds both at startup).

---

tl;dr: SteamVR and Oculus' runtime don't support switching between each other. VRTK's ability to change the used SDK at runtime can't work around that.

---

BTW: The SDK Switcher is just a GUI - this issue is more about VRTK's runtime SDK switching support, but that's nitpicking 😄

**Comment by guiglass** [comment]
Hi bddckr, thanks for that explanation. Makes me a little uncertain now on how to create one game for both HMDs, I am new to VR development, and just got a Rift literally a few days ago.

I was hoping to find a fast approach to development where I could use the Vive mostly, but occasionally switch over to the Oculus and just pick up right where the Vive left off (like switching to the simulator).

Up till now I always thought the SDK switcher was capable of doing that.
If this is not possible, what would be recommended? Because at present, I have to do literally 10 steps including a full Unity restart just to switch over to another HMD! 

-----------------
From what I understood, the SteamVR runtime should be capable of handling both headsets, so only that would need to be restarted?

I looked through some Google searches for ways to configure OpenVR to favor the Oculus over the Vive and use it withing SteamVR. I was able to unplug the Vive and that got it going. But I tried things like "C:\Program Files (x86)\Steam\steamapps\common\SteamVR\bin\win64\vrstartup.exe - oculus" which had no effect. Any hints?

-----------------
Just a little side question. If I plan to release a game (on steam) would using the pure SteamVR method for both headsets be recommended? Or should I attempt to implement OculusVR at the very end??

Thanks!!!

**Comment by bddckr** [comment]
Stop switching at _runtime_ between both headsets is my recommendation. You can easily test both otherwise. It's just not easy to switch between SteamVR and Oculus with a _running_ VR app.

There's no need to test both devices all the time during development - especially with VRTK since it abstracts everything for you to have one shared API layer to work with, without bothering with the SDK and device differences much.

You just have to get the input button settings right at one point, and ensure stuff like properly adjusted snap handles per supported controller etc.

---

AFAIK it's possible to force a driver by using `forcedDriver` in the config, probably only picked up after restarting SteamVR: https://developer.valvesoftware.com/wiki/SteamVR/steamvr.vrsettings

---

VRTK allows you to have one build that supports everything and given that Rift users want Oculus Home (i.e. Oculus SDK) support even when buying a title on Steam you should do what all VRTK example scenes do: Have all your supported SDK Setups in and ensure the order works.

The example scenes put the Oculus before the SteamVR SDK Setup in the SDK Manager to allow Rift users to always get the native support first. If they want to run via SteamVR instead you can just offer an additional launch option: `-vrmode openvr` tries to load and use the SteamVR SDK Setup first instead.
(Steam offers launch parameter settings when you set up your game on their store - that's what you probably saw by launching some titles on the desktop and getting a window to select which mode to start, or when right clicking on the Steam library entry of the VR title in question and then using one of the launch menu items right at the top.)

**Comment by guiglass** [comment]
@bddckr 
Thank so much for the info, I think I can get the hang of using both headsets in a project now.

 ------------------------

> AFAIK it's possible to force a driver by using forcedDriver in the config

I wrote a script for handling the switching of the "forcedDriver" parameter from within Unity on Windows. This does save some time (not having to locate and edit the txt file manually), however Unity still requires a restart, otherwise SteamVR says "Unresponsive Unity.exe" regardless which headset was just configured.

SteamVR is completely restarted and successfully picks up the desired HMD, but maybe there is something in Unity that needs to be reset too?

just add SteamVRDriverSwitcher.cs component to an empty gameobject in a scene, select desired driver and press the Load Driver button.

**SteamVRDriverSwitcher.cs**
`using UnityEngine;
using System;
using System.Diagnostics;
using System.Text.RegularExpressions;
using UnityEditor;

[ExecuteInEditMode]
public class SteamVRDriverSwitcher : MonoBehaviour {
	
	public enum VRDriver {
		Vive,
		Rift
	}
	[SerializeField] VRDriver driver;

	[Tooltip ("Path to the steamvr.vrsettings file that lives in the /steam/config directory")]
	[SerializeField] string steamvrSettingsPath = "C:\\Program Files (x86)\\Steam\\config\\steamvr.vrsettings"; 
	[Tooltip ("Path to the vrstartup.exe file for launching SteamVR")]
	[SerializeField] string vrstartupPath = "C:\\Program Files (x86)\\Steam\\steamapps\\common\\SteamVR\\bin\\win64\\vrstartup.exe";

	void LoadExternalJsonfile(string filePath) {
		string jsonString = System.IO.File.ReadAllText(filePath);

		var match = Regex.Match (jsonString, @"""steamvr"".*:.*{([^}]*)}").Groups [1];
		int start = match.Index;
		match = Regex.Match (match.Value, @"""forcedDriver"".*:.*""([^""].*)""").Groups [1];
		start += match.Index;

		string newJsonString = jsonString.Remove (start, match.Value.Length);
		switch (driver) {
		case VRDriver.Vive :
			newJsonString = newJsonString.Insert(start, "lighthouse");
			break;
		case VRDriver.Rift :
			newJsonString = newJsonString.Insert(start, "oculus");
			break;
		}
		System.IO.File.WriteAllText (filePath, newJsonString);
	}

	public void LoadSelectedDriver () {
		try {
			Process shellProcess = new Process();
			shellProcess.StartInfo.WindowStyle = ProcessWindowStyle.Hidden;
			shellProcess.StartInfo.CreateNoWindow = true;
			shellProcess.StartInfo.UseShellExecute = false;
			shellProcess.StartInfo.FileName = "C:\\Windows\\system32\\cmd.exe";
			shellProcess.StartInfo.Arguments = "/c taskkill /F /IM \"vrmonitor.exe\" /T";
			shellProcess.EnableRaisingEvents = true;
			shellProcess.Start();
			shellProcess.WaitForExit();
			int ExitCode = shellProcess.ExitCode;

			LoadExternalJsonfile(steamvrSettingsPath);

		} catch (Exception e){
			print(e);        
		}
	}

	public void StartSteamVR () {
		Process vrmonitorProcess = new Process();
		vrmonitorProcess.StartInfo.FileName = vrstartupPath;
		vrmonitorProcess.Start();
	}
}`

**SteamVRDriverSwitcherEditor.cs**
`using UnityEngine;
using System.Collections;
using UnityEditor;

[CustomEditor(typeof(SteamVRDriverSwitcher))]
public class SteamVRDriverSwitcherEditor : Editor
{
	public override void OnInspectorGUI()
	{
		DrawDefaultInspector();
		SteamVRDriverSwitcher myScript = (SteamVRDriverSwitcher)target;
		GUILayout.Space (10);
		GUILayout.Label("Close SteamVR and change the driver configuration.");
		if (GUILayout.Button ("Load Driver")) {
			myScript.LoadSelectedDriver ();
		}
		GUILayout.Space (10);
		GUILayout.Label("Wait a few seconds prior to re-opening!");
		if (GUILayout.Button ("Start SteamVR")) {
			myScript.StartSteamVR ();
		}

	}
}`

**Comment by guiglass** [comment]
I notice some unexpected behavior coming from VRTK_SDKManager that causes the Oculus not to start sometimes. It has to do with the way the list of virtual reality SDK under Player > XR settings are automatically managed. 

For some reason, even if I remove the "None" option and then open any of the VRTK examples, the "None" option has once again been added to the list (and is always placed on top).

Whenever the "None" option exists, the Oculus fails to start!

The solution I am using at the moment is to edit the file ProjectSettings>ProjectSettings.asset and remove the "- None" option prior to opening Unity.
```
  m_BuildTargetVRSettings:
  - m_BuildTarget: Standalone
    m_Enabled: 1
    m_Devices:
    - None //THIS MUST BE REMOVED MANUALLY
    - Oculus
    - OpenVR
```
I don't think the "auto manage vr settings" is going to help either because if any other scene has it enabled then the None option is immediately added again and get stuck in there until I remove it again.
Could this be a bug? Or is there a reason that "None" must always be present and on top in the SDK list?

**Comment by bddckr** [comment]
It's always added there if you have more than one SDK Setup on your SDK Manager to allow the dynamic loading of one after the other at startup of your app. This is what makes the behavior of "first try Oculus SDK, but fall back to SteamVR if not possible" possible in the first place.

**Comment by Zenopheus** [comment]
I just wasted an half hour figuring this out. "None" kept getting added to my XR Settings panel. Then I thought "VRTK usually makes broad assumptions like this." Sure enough! I have some scenes that don't use VRTK and this causes more problems than it's worth. This kind of stuff is why we have chosen not to employ VRTK in all of our scenes. This issue just pissed off my team and made VRTK even worse in their eyes. **Please remove this magic feature.**

**Comment by bddckr** [comment]
@Zenopheus Hey there! So sorry you're having a hard time, this is definitely not a goal of our effort here!

As you may not be aware there's a checkbox on your SDK Manager to control this. Note that you have to have None in the XR list to make runtime switching to/from Simulator work as well as supporting automatically selecting the proper SDK (in order you specify). If you don't get value out of either feature I'm sorry that the SDK Manager's default is to handle it for you. Thankfully it's an option, though!

I suggest joining our [![Slack](http://sysdia2.co.uk/badge.svg)](http://invite.vrtk.io) to prevent you from taking longer to find the documentation of the components you're using which ultimately should speed up your development process. Our community is always helping out each other!

v4 of VRTK doesn't even have an SDK Manager so we're ahead of your request.

---

Should you want to discuss any of this please make sure to do so in the Slack because GitHub issues are better for bug reports.

**Comment by felipepedroso** [comment]
Hi @bddckr, the same behavior is happening here and the application never appears on the Rift. I tried to find the option that you mentioned, but it seems that it doesn't exist on the latest version. Do you know how to fix that:
